properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+        * You may also specify a **registry** - This is the flat dictionary
+          where UID/HasUID pairs are stored. By default, no registry need
+          be provided; a new dictionary will be created.
+        * Any other keyword arguments will be passed through to the Property
+          serializers.
+        """
+        registry = kwargs.pop('registry', None)
+        if registry is None:
+            registry = dict()
+        if not registry:
+            root = True
+            registry.update({'__uid__': self.uid})
+        else:
+            root = False
+        key = self.uid
+        if key not in registry:
+            registry.update({key: None})
+            registry.update({key: super(HasUID, self).serialize(
+                registry=registry,
+                include_class=include_class,
+                save_dynamic=save_dynamic,
+                **kwargs
+            )})
+        if root:
+            return registry
+        return key
+
+    @classmethod
+    def deserialize(cls, value, trusted=False, strict=False,
+                    assert_valid=False, **kwargs):
+        """Deserialize nested HasUID instance from flat pointer dictionary
+
+        **Parameters**
+
+        * **value** - Flat pointer dictionary produced by :code:`serialize`
+          with UID/HasUID key/value pairs. It also includes a
+          :code:`__uid__` key to specify the root HasUID instance.
+        * **trusted** - If True (and if the input dictionaries have
+          :code:`'__class__'` keyword and this class is in the registry), the
+          new **HasProperties** class will come from the dictionary.
+          If False (the default), only the **HasProperties** class this
+          method is called on will be constructed.
+        * **strict** - Requires :code:`'__class__'`, if present on the input
+          dictionary, to match the deserialized instance's class. Also
+          disallows unused properties in the input dictionary. Default
+          is False.
+        * **assert_valid** - Require deserialized instance to be valid.
+          Default is False.
+        * You may also specify an alternative **uid** - This allows a different
+          HasUID root instance to be specified. It overrides :code:`__uid__`
+          in the input dictionary.
+        * Any other keyword arguments will be passed through to the Property
+          deserializers.
+
+        .. note::
+
+            UidModel instances are constructed with no input arguments
+            (ie :code:`cls()` is called). This means deserialization will
+            fail if the init method has been overridden to require
+            input parameters.
+        """
+        registry = kwargs.pop('registry', None)
+        if registry is None:
+            if not isinstance(value, dict):
+                raise ValueError('HasUID must deserialize from dictionary')
+            registry = value.copy()
+            uid = kwargs.get('uid', registry.get('__uid__'))
```

Comment:
Okay, I sort of see the naming here. Still wonder if something like `__root__` is more descriptive?